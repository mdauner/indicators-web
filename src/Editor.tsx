import React from 'react';
import { DataSet } from './types';

interface Props {
  selectedDataSet?: DataSet;
  selectedYear?: string;
  value?: string;
  onChange: (value: string) => void;
  onSave: () => void;
}

export default function Editor({
  selectedDataSet,
  selectedYear,
  value,
  onChange,
  onSave
}: Props) {
  if (!selectedDataSet || !selectedYear) {
    return null;
  }

  return (
    <div data-testid="editor" className="mt-8 flex items-center">
      <div>
        Update value of {selectedDataSet.country} ({selectedYear}):
      </div>
      <input
        data-testid="input"
        type="number"
        className="ml-4 bg-white border border-gray-300 rounded-lg py-2 px-4 block appearance-none leading-normal"
        value={value}
        onChange={(e) => onChange(e.target.value)}
      />
      <button
        onClick={onSave}
        disabled={!value || isNaN(parseFloat(value))}
        className="ml-4 bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded disabled:opacity-50 disabled:cursor-not-allowed"
      >
        Save
      </button>
    </div>
  );
}
