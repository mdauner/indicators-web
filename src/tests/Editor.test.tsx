import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Editor from '../Editor';
import { mockDataSet } from './mocks';

describe('Editor', () => {
  test('returns null when no cell is selected', async () => {
    const { container } = render(
      <Editor onChange={() => true} onSave={() => true} />
    );

    expect(container.firstChild).toBeNull();
  });

  test('renders component', async () => {
    const selectedYear = '1961';
    const { container } = render(
      <Editor
        selectedDataSet={mockDataSet}
        selectedYear={selectedYear}
        value="125.2"
        onChange={() => true}
        onSave={() => true}
      />
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  test('calls onChange callback when input text is changed', async () => {
    const selectedYear = '1961';
    const newValue = '111';
    const onChange = jest.fn();

    const { getByTestId } = render(
      <Editor
        selectedDataSet={mockDataSet}
        selectedYear={selectedYear}
        value="125.2"
        onChange={onChange}
        onSave={() => true}
      />
    );
    const input = getByTestId('input');
    fireEvent.change(input, { target: { value: newValue } });

    expect(onChange).toHaveBeenCalledWith(newValue);
  });

  test('calls onSave callback when save button is clicked', async () => {
    const selectedYear = '1961';
    const onSave = jest.fn();

    const { getByText } = render(
      <Editor
        selectedDataSet={mockDataSet}
        selectedYear={selectedYear}
        value="123.1"
        onChange={() => true}
        onSave={onSave}
      />
    );
    const saveButton = getByText('Save');

    expect(saveButton).not.toBeDisabled();

    fireEvent.click(saveButton);
    expect(onSave).toHaveBeenCalled();
  });

  test('disables save button when value is invalid', async () => {
    const selectedYear = '1961';
    const onSave = jest.fn();

    const { getByText } = render(
      <Editor
        selectedDataSet={mockDataSet}
        selectedYear={selectedYear}
        value="invalid_value"
        onChange={() => true}
        onSave={onSave}
      />
    );
    const saveButton = getByText('Save');

    expect(saveButton).toBeDisabled();
  });
});
