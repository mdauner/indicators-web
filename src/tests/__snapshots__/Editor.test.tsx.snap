// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editor renders component 1`] = `
<div
  class="mt-8 flex items-center"
  data-testid="editor"
>
  <div>
    Update value of 
    USA
     (
    1961
    ):
  </div>
  <input
    class="ml-4 bg-white border border-gray-300 rounded-lg py-2 px-4 block appearance-none leading-normal"
    data-testid="input"
    type="number"
    value="125.2"
  />
  <button
    class="ml-4 bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded disabled:opacity-50 disabled:cursor-not-allowed"
  >
    Save
  </button>
</div>
`;
